gadget/install/install.go

```diff
@@ -377,6 +376,141 @@ func Run(model gadget.Model, gadgetRoot, kernelRoot, bootDevice string, options
+		// We remove /dev from p.Node here
+		if filepath.Base(p.Node) == partDevice {
+			return &p, nil
+		}
+	}
+
+	return nil, fmt.Errorf("cannot find partition %q", partDevice)
+}
+
+// laidOutStructureForDiskStructure searches for the laid out structure that
+// matches a given OnDiskStructure.
+func laidOutStructureForDiskStructure(laidVols map[string]*gadget.LaidOutVolume, gadgetVolName string, onDiskStruct *gadget.OnDiskStructure) (
+	*gadget.LaidOutStructure, error) {
+
+	for _, laidVol := range laidVols {
+		// Check that this is the right volume
+		if laidVol.Name != gadgetVolName {
+			continue
+		}
+		for _, laidStruct := range laidVol.LaidOutStructure {
+			if onDiskStruct.Name == laidStruct.Name {
+				return &laidStruct, nil
+			}
+		}
+	}
+
+	return nil, fmt.Errorf("cannot find laid out structure for %q", onDiskStruct.Name)
+}
+
+// sysfsPathForBlockDevice returns the sysfs path for a block device.
+func sysfsPathForBlockDevice(device string) (string, error) {
+	partPath, err := os.Readlink(filepath.Join("/sys/class/block", device))
+	if err != nil {
+		return "", err
+	}
+	// Remove initial ../../ from partPath, and make path absolute
+	return filepath.Join("/sys/class/block", partPath), nil
+}
+
+// onDiskVolumeFromPartitionSysfsPath creates an OnDiskVolume that
+// matches the disk that contains the given partition sysfs path
+func onDiskVolumeFromPartitionSysfsPath(partPath string) (*gadget.OnDiskVolume, error) {
+	// Removing the last component will give us the disk path
+	diskPath := filepath.Dir(partPath)
+	disk, err := disks.DiskFromDevicePath(diskPath)
+	if err != nil {
+		return nil, fmt.Errorf("cannot retrieve disk information for %q: %v", partPath, err)
+	}
+	onDiskVol, err := gadget.OnDiskVolumeFromDisk(disk)
+	if err != nil {
+		return nil, fmt.Errorf("cannot retrieve on disk volume for %q: %v", partPath, err)
+	}
+
+	return onDiskVol, nil
+}
+
+// applyLayoutToOnDiskStructure finds the on disk structure from a
+// sysfs partition path string and takes the laid out information from
+// laidOutVols and inserts it there.
+func applyLayoutToOnDiskStructure(onDiskVol *gadget.OnDiskVolume, partPath string, laidOutVols map[string]*gadget.LaidOutVolume, gadgetVolName string) (*gadget.OnDiskStructure, error) {
+	partDevice := filepath.Base(partPath)
+	onDiskStruct, err := structureFromPartDevice(onDiskVol, partDevice)
+	if err != nil {
+		return nil, fmt.Errorf("cannot find partition %q: %v", partPath, err)
+	}
+
+	laidOutStruct, err := laidOutStructureForDiskStructure(laidOutVols, gadgetVolName, onDiskStruct)
+	if err != nil {
+		return nil, err
+	}
+	// This fills LaidOutStructure, including (importantly) the ResolvedContent field
+	onDiskStruct.LaidOutStructure = *laidOutStruct
+
+	return onDiskStruct, nil
+}
+
+// WriteContent writes gadget content to the devices specified in
+// onVolumes. It returns the resolved on disk volumes.
+func WriteContent(onVolumes map[string]*gadget.Volume, observer gadget.ContentObserver,
```

Comment:
do we need a unit test for this?

Reply:
I've started with this but I think it will be better to have as a follow up if that's ok, as it involves quite a bit of mocking because of disk accesses.

Reply:
Added